[14.0][IMP] stachechart_mixin: make it possible to ignore some events in the has_allowed_event compute

jdoutreloux opened the pull request:
This commits makes it possible to ignore some events when computing has_allowed_events by passing them in the context key `ignore_for_has_allowed_events`.
For example, our record has a cancel action on each state but we don't want to take that action into account when getting the records with allowed_events
Our use case is showing all the records that need an action in a dashboard. But the records with a cancel action are not actually in need of an action.

sbejaoui commented on statechart/models/statechart_mixin.py:
```diff
@@ -321,11 +321,49 @@ def _get_sc_event_allowed_field_names(self):
+        """
+        self.ensure_one()
+        allowed_transitions = []
+        for event in allowed_events:
+            # The states are ordered by depth in the interpreter's configuration.
+            # The last one is our actual state.
+            state = self.sc_interpreter.configuration[-1]
+            state_transitions = self._statechart.transitions_from(state)
+            for transition in state_transitions:
+                if transition.event == event:
+                    for priority in priorities:
+                        if transition.priority == priority:
+                            allowed_transitions += [transition]
+                    if not priorities:
+                        allowed_transitions += [transition]
```
```suggestion
        state = self.sc_interpreter.configuration[-1]
        state_transitions = self._statechart.transitions_from(state)
        for transition in state_transitions:
            if transition.event in allowed_events and (not priorities or transition.priority in priorities):
                allowed_transitions.append(transition)
```

sbejaoui commented on statechart/models/statechart_mixin.py:
```diff
@@ -321,11 +321,49 @@ def _get_sc_event_allowed_field_names(self):
+
     @api.depends("sc_state")
     def _compute_sc_has_allowed_events(self):
+        priorities = self.env.context.get("priorities")
```
```suggestion
        priorities = self.env.context.get("priorities", [])
```

sbejaoui commented on statechart/models/statechart_mixin.py:
```diff
@@ -321,11 +321,49 @@ def _get_sc_event_allowed_field_names(self):
-            rec.sc_has_allowed_events = any([rec[f] for f in sc_fields])
+            events = rec._get_sc_allowed_events(sc_fields)
+            if not priorities:
+                rec.sc_has_allowed_events = True if events else False
```
```suggestion
                rec.sc_has_allowed_events = bool(events)
```

sbejaoui reviewed (changes requested):
It's a somewhat tricky solution. I didn't like the fact that it depends on the context. 

As I understand it, we are trying to exclude certain events from the user notification/dashboard. I prefer a more explicit field than the priority. For example: "exclude_from_allowed_event." takes true or false.

jdoutreloux committed:
[IMP] stachechat_mixin: make it possible to ignore some events in the has_allowed_event compute

sbidoul commented on statechart/models/statechart_mixin.py:
```diff
@@ -316,7 +316,8 @@ class can override the statechart with another one adding new events,
     def _get_sc_event_allowed_field_names(self):
-        event_names = self._statechart.events_for()
+        ignore_for_has_allowed_events = self.env.context.get("ignore_for_has_allowed_events", [])
+        event_names = [name for name in self._statechart.events_for() if name not in ignore_for_has_allowed_events]
```
nit: you can put the `if name not in ignore_for_has_allowed_events` in the `return` statement, to avoid creating a temporary list.

sbidoul commented on statechart/models/statechart_mixin.py:
```diff
@@ -316,7 +316,8 @@ class can override the statechart with another one adding new events,
-        event_names = self._statechart.events_for()
+        ignore_for_has_allowed_events = self.env.context.get("ignore_for_has_allowed_events", [])
+        event_names = [name for name in self._statechart.events_for() if name not in ignore_for_has_allowed_events]
         return [
             _sc_make_event_allowed_field_name(event_name) for event_name in event_names
         ]
```
```suggestion
        return [
            _sc_make_event_allowed_field_name(event_name) for event_name in event_names if event_name not in ignore_for_has_allowed_event
        ]
```

sbejaoui reviewed (approved):
LGTM

sbidoul commented:
This needs a test. It should be easy to add in `test_basic.py`. Don't hesitate to add events to `test_statechart.yml` to make the test more relevant.

jdoutreloux commented:
> This needs a test. It should be easy to add in `test_basic.py`. Don't hesitate to add events to `test_statechart.yml` to make the test more relevant.

Sorry for the delay, I'm just now seeing your review. I've added the test as you asked.

sbidoul committed:
[IMP] handle ignore_for_has_allowed_events only in the compute method

There is no need to leak this context handling to
_get_sc_event_allowed_field_names, where
this can be handled with a new argument.
